service/qos/raft_service_level_distributed_data_accessor.cc

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "raft_service_level_distributed_data_accessor.hh"
+#include "cql3/query_processor.hh"
+#include "service/qos/qos_table.hh"
+#include "service/raft/raft_group0_client.hh"
+#include "qos_table.hh"
+#include "db/system_keyspace.hh"
+#include "types/types.hh"
+
+namespace qos {
+
+raft_service_level_distributed_data_accessor::raft_service_level_distributed_data_accessor(cql3::query_processor& qp, service::raft_group0_client& group0_client)
+    : _qp(qp)
+    , _group0_client(group0_client) {}
+
+future<qos::service_levels_info> raft_service_level_distributed_data_accessor::get_service_levels() const {
+    return qos::get_service_levels(_qp, db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2);
+}
+
+future<qos::service_levels_info> raft_service_level_distributed_data_accessor::get_service_level(sstring service_level_name) const {
+    return qos::get_service_level(_qp, db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2, service_level_name);
+}
+
+future<> raft_service_level_distributed_data_accessor::start_service_level_operation() {
+    _group0_guard = co_await _group0_client.start_operation(&_aborter);
+}
+
+void raft_service_level_distributed_data_accessor::abort_service_level_operation() {
+    if (_group0_guard) {
+        _aborter.request_abort();
+    }
+    _group0_guard = std::nullopt;
+}
+
+future<> raft_service_level_distributed_data_accessor::do_raft_command(std::function<future<std::vector<mutation>>(api::timestamp_type)> mutation_function, sstring description) {
+    if (!_group0_guard) {
+        throw runtime_exception("raft_service_level_distributed_data_accessor: operation has to be started before doing any changes");
+    }
+
+    auto timestamp = _group0_guard->write_timestamp();
+
+    auto muts = co_await mutation_function(timestamp);
+    service::write_mutations change {
+        .mutations{muts.begin(), muts.end()},
+    };
+
+    service::group0_guard& g = *_group0_guard;
+    auto group0_cmd = _group0_client.prepare_command(change, g, description);
+    co_await _group0_client.add_entry(std::move(group0_cmd), std::move(*_group0_guard), &_aborter);
+    _group0_guard = std::nullopt;
```

Comment:
If the caller has an exception, this won't be executed, and Raft will stall.

It's dangerous to store a transient variable in a persistent object, better to obtain a transaction object. It will require changing the data accessor interface, but it's better than debugging stalled raft.

Reply:
Changed the interface to use a transactions

Reply:
Now changed to use `group0_guard` from cql statement and it's passed as a parameter to the function